### ISSUE TITLE
Feature: lingui extract should collect messages from multiple sources

### ISSUE THREAD

**Issue by HamidAghdaee** [lingui/js-lingui#257]
Thank you so much for this wonderful library. 

One issue we're facing at the moment::

Our application is large enough that we want to extract translations from various src folders in our application into different locale folders. Currently there is no way to pass certain config options to `lingui extract` command such as `srcPathDirs` and `<rootDir>/locale`

It would be nice to either be able to pass these in as parameters or to force the command to read these from a custom config file.

Thanks again.

**Comment by tricoder42** [comment]
Hey @HamidAghdaee,
I was facing similar issue some time ago but was lazy enough to find a temporary workaround :D I'm working on CLI atm, so let's focus on this.

Just to clarify: you have N source folders and you would like to extract message to N separate locale folders, right?

Something like:
```json
{
   "localeDirs": {
      "./locale/package1": "./src/package1",
      "./locale/package2": "./src/package2",
    }
}
```

**Comment by HamidAghdaee** [comment]
@tricoder42 yes. that would solve part of our problem. Our use case is more like what you're trying to solve here: https://github.com/lingui/js-lingui/issues/14

We maintain our code in packages in a mono-repo of npm published packages, manages by yarn workspaces and Lerna. 

We want each package to maintain their own local locale files. For one of the packages there are enough translations that having more than one locale folder is a necessity. For the others one locale folder is sufficient, but I don't want to install lingui for each package. I want to run `lingui extract` as part of a `lerna exec` command, which should pass the  local config file to the package to the lingui command. 

Let me know what you think.

Thanks.

**Comment by tricoder42** [comment]
Make sense. It should be enough to have `@lingui/cli` installed in the root of monorepo and then running `lingui add-locale|extract|compile` in root would execute the command in each workspace.

That's what I was suggesting with
```json
{
   "localeDirs": {
      "./locale/package1": "./src/package1",
      "./locale/package2": "./src/package2",
    }
}
```

but we could pick this config from `workspaces` config.

Regarding namespaces: This might be indeed related, but `namespaces` are mostly about having a single `locale` folder which collects messages from several other sources (1:N mapping). For example you might have packages `App`, `CreditCardInput`, `RelativeDateTime` and messages from all of them will be collected in the same locale directory, but msgIds from `CreditCardInput` will be namespaced with `@CreditCardInput/` and so on:

```json
{
   "msg.header": "Header from App", 
   "@CreditCardInput/msg.header": "Header from CreditCardInput",
   "@RelativeDateTime/msg.header": "Header from RelativeDateTime",
}
```

**Comment by HamidAghdaee** [comment]
@tricoder42 with regards to your first point:

I just don't to have to add a row for each package to the config you mentioned. `lerna exec` already executes any command for every package. And I want to direct it to use the config file local to the package for which it is executing. 

I agree with your point about name-spacing. That could be helpful for us. One challenge I see however is:

1. If the messageId is extracted to be the same as message value, then how do you namespace it? Would there be a namespacing not based on the key, but something else? I'm talking about the case where the developer doesn't provide an explicit id.

thanks.

**Comment by tricoder42** [comment]
> I just don't to have to add a row for each package to the config you mentioned. lerna exec already executes any command for every package. And I want to direct it to use the config file local to the package for which it is executing.

I understand that.

There' one possible workaround which should work out of the box: Add lingui config to each package:

```json
{
  "lingui": {
    "localeDir": "./locale"
  }
}
```

The problem is that if you have any additional config in root, it will be overwritten. On the other hand, this should create `locale` dir inside each package with such config.

Namespacing is still in progress and there're many challenges to solve. But yes, if developer doen't provide explicit ID, then the ID is the message itself. Instead of `@namespace/msg.id` it would be `@namespace/Message string`. There's no difference. The messageID isn't visible for user anyway, so it could be anything.

**Comment by HamidAghdaee** [comment]
I see. You're right about the packages within the monorepo solution you mentioned. I forgot that lerna exec makes the cwd the package in which it's executing. Having multiple src and locale combos in the same config is still useful for us for the case of splitting translations in the same package. 

With regards to name-spacing:

How do you intend to determine what the namespace is to be if not specified in the key by the user? That's what i18next does and I have to say I don't love it. 

On a related note, it would be nice to be able to do some mapping of keys when the keys are  generated messageIds. After all, for the English (in our case our default lanaguage) you end up with this in your en.json file:

{
  "long message....": "long message"
}

It would be nice to make it so that a mapping function can be applied generate keys from the value (or other factors such as file name, etc) to shorten the key string.

**Comment by tricoder42** [comment]
I haven't really thought about all edge cases of namespacing. I'm working on #179, adding support for .po and possibly .xliff files. Also I'm preparing a video tutorial, so there's too many things going on in my head and I need to release 3.0 anyway before namespaces :) If you have any good points or suggestions, feel free to share it in #14, but right now I'll I can give you is my idea of solution, which is far from the final solution :)

**Comment by tricoder42** [comment]
Here's a [draft](https://lingui.js.org/rfc/002_locale_dirs.html#considered-use-cases) of splitting source catalogs. Any comments welcome.

**Comment by tricoder42** [comment]
I'm planning to work & release it next weekend. It's been on a paper for too long.

At the moment there's no way to have more than one message catalog, unfortunately.

**Comment by tricoder42** [comment]
New catalogs were already merged to `next` branch and will be released in next major release, LinguiJS v3.0. See #334 how to test pre-release.
